[Flaking Test] [sig-api-machinery] k8s.io/kubernetes/test/integration/apiserver/coordinatedleaderelection.coordinatedleaderelection

### Which jobs are flaking?

master-blocking
- integration-master

### Which tests are flaking?

k8s.io/kubernetes/test/integration/apiserver/coordinatedleaderelection.coordinatedleaderelection
[Prow](https://prow.k8s.io/view/gs/kubernetes-ci-logs/logs/ci-kubernetes-integration-master/1881787128904421376)
[Triage](https://storage.googleapis.com/k8s-triage/index.html?test=k8s.io%2Fkubernetes%2Ftest%2Fintegration%2Fapiserver%2Fcoordinatedleaderelection.coordinatedleaderelection&xjob=e2e-kops)

### Since when has it been flaking?

[1/14/2025, 12:41:26 AM](https://prow.k8s.io/view/gs/kubernetes-ci-logs/logs/ci-kubernetes-integration-master/1878882224963588096)
[1/18/2025, 1:29:23 AM](https://prow.k8s.io/view/gs/kubernetes-ci-logs/logs/ci-kubernetes-integration-master/1880343855438499840)
[1/20/2025, 1:20:29 PM](https://prow.k8s.io/view/gs/kubernetes-ci-logs/logs/ci-kubernetes-integration-master/1881247568734720000)
[1/20/2025, 1:37:27 PM](https://prow.k8s.io/view/gs/kubernetes-ci-logs/logs/ci-kubernetes-integration-1-32/1881251846668947456)
[1/22/2025, 1:04:24 AM](https://prow.k8s.io/view/gs/kubernetes-ci-logs/logs/ci-kubernetes-integration-master/1881787128904421376)
[1/23/2025, 1:45:57 PM](https://prow.k8s.io/view/gs/kubernetes-ci-logs/logs/ci-kubernetes-integration-1-32/1882341172085526528)

### Testgrid link

https://testgrid.k8s.io/sig-release-master-blocking#integration-master

### Reason for failure (if possible)

```
{Failed  === RUN   TestCoordinatedLeaderElectionLeaseTransfer
    testserver.go:582: Resolved testserver package path to: "/home/prow/go/src/k8s.io/kubernetes/cmd/kube-apiserver/app/testing"
    testserver.go:402: runtime-config=map[api/all:true]
    testserver.go:403: Starting kube-apiserver on port 43809...
    testserver.go:438: Waiting for /healthz to be ok...
[-]poststarthook/start-apiextensions-controllers failed: not finished
[-]poststarthook/crd-informer-synced failed: not finished
[-]poststarthook/start-service-ip-repair-controllers failed: not finished
[-]poststarthook/scheduling/bootstrap-system-priority-classes failed: not finished
[-]poststarthook/priority-and-fairness-config-producer failed: not finished
[-]poststarthook/bootstrap-controller failed: not finished
[-]poststarthook/apiservice-registration-controller failed: not finished
[-]poststarthook/apiservice-discovery-controller failed: not finished
[-]autoregister-completion failed: missing APIService: [v1. v1.admissionregistration.k8s.io v1.apiextensions.k8s.io v1.apps v1.authentication.k8s.io v1.authorization.k8s.io v1.autoscaling v1.batch v1.certificates.k8s.io v1.coordination.k8s.io v1.discovery.k8s.io v1.events.k8s.io v1.flowcontrol.apiserver.k8s.io v1.networking.k8s.io v1.node.k8s.io v1.policy v1.rbac.authorization.k8s.io v1.scheduling.k8s.io v1.storage.k8s.io v1alpha1.admissionregistration.k8s.io v1alpha1.internal.apiserver.k8s.io v1alpha1.storage.k8s.io v1alpha2.coordination.k8s.io v1alpha3.resource.k8s.io v1beta1.admissionregistration.k8s.io v1beta1.networking.k8s.io v1beta1.resource.k8s.io v1beta1.storage.k8s.io v2.autoscaling]
[-]poststarthook/start-apiextensions-controllers failed: not finished
[-]poststarthook/crd-informer-synced failed: not finished
[-]poststarthook/scheduling/bootstrap-system-priority-classes failed: not finished
[-]poststarthook/priority-and-fairness-config-producer failed: not finished
[-]poststarthook/bootstrap-controller failed: not finished
[-]poststarthook/apiservice-registration-controller failed: not finished
[-]poststarthook/apiservice-discovery-controller failed: not finished
[-]autoregister-completion failed: missing APIService: [v1. v1.admissionregistration.k8s.io v1.apiextensions.k8s.io v1.apps v1.authentication.k8s.io v1.authorization.k8s.io v1.autoscaling v1.batch v1.certificates.k8s.io v1.coordination.k8s.io v1.discovery.k8s.io v1.events.k8s.io v1.flowcontrol.apiserver.k8s.io v1.networking.k8s.io v1.node.k8s.io v1.policy v1.rbac.authorization.k8s.io v1.scheduling.k8s.io v1.storage.k8s.io v1alpha1.admissionregistration.k8s.io v1alpha1.internal.apiserver.k8s.io v1alpha1.storage.k8s.io v1alpha2.coordination.k8s.io v1alpha3.resource.k8s.io v1beta1.admissionregistration.k8s.io v1beta1.networking.k8s.io v1beta1.resource.k8s.io v1beta1.storage.k8s.io v2.autoscaling]
[-]poststarthook/start-apiextensions-controllers failed: not finished
[-]poststarthook/crd-informer-synced failed: not finished
[-]poststarthook/scheduling/bootstrap-system-priority-classes failed: not finished
[-]poststarthook/priority-and-fairness-config-producer failed: not finished
[-]autoregister-completion failed: missing APIService: [v1. v1.admissionregistration.k8s.io v1.apiextensions.k8s.io v1.apps v1.authentication.k8s.io v1.authorization.k8s.io v1.autoscaling v1.batch v1.certificates.k8s.io v1.coordination.k8s.io v1.discovery.k8s.io v1.events.k8s.io v1.flowcontrol.apiserver.k8s.io v1.networking.k8s.io v1.node.k8s.io v1.policy v1.rbac.authorization.k8s.io v1.scheduling.k8s.io v1.storage.k8s.io v1alpha1.admissionregistration.k8s.io v1alpha1.internal.apiserver.k8s.io v1alpha1.storage.k8s.io v1alpha2.coordination.k8s.io v1alpha3.resource.k8s.io v1beta1.admissionregistration.k8s.io v1beta1.networking.k8s.io v1beta1.resource.k8s.io v1beta1.storage.k8s.io v2.autoscaling]
[-]poststarthook/scheduling/bootstrap-system-priority-classes failed: not finished
[-]poststarthook/priority-and-fairness-config-producer failed: not finished
[-]autoregister-completion failed: missing APIService: [v1. v1.admissionregistration.k8s.io v1.apiextensions.k8s.io v1.apps v1.authentication.k8s.io v1.authorization.k8s.io v1.autoscaling v1.batch v1.certificates.k8s.io v1.coordination.k8s.io v1.discovery.k8s.io v1.events.k8s.io v1.flowcontrol.apiserver.k8s.io v1.networking.k8s.io v1.node.k8s.io v1.policy v1.rbac.authorization.k8s.io v1.scheduling.k8s.io v1.storage.k8s.io v1alpha1.admissionregistration.k8s.io v1alpha1.internal.apiserver.k8s.io v1alpha1.storage.k8s.io v1alpha2.coordination.k8s.io v1alpha3.resource.k8s.io v1beta1.admissionregistration.k8s.io v1beta1.networking.k8s.io v1beta1.resource.k8s.io v1beta1.storage.k8s.io v2.autoscaling]
[-]poststarthook/scheduling/bootstrap-system-priority-classes failed: not finished
[-]poststarthook/priority-and-fairness-config-producer failed: not finished
[-]autoregister-completion failed: missing APIService: [v1. v1.admissionregistration.k8s.io v1.apiextensions.k8s.io v1.apps v1.authentication.k8s.io v1.authorization.k8s.io v1.autoscaling v1.batch v1.certificates.k8s.io v1.coordination.k8s.io v1.discovery.k8s.io v1.events.k8s.io v1.flowcontrol.apiserver.k8s.io v1.networking.k8s.io v1.node.k8s.io v1.policy v1.rbac.authorization.k8s.io v1.scheduling.k8s.io v1.storage.k8s.io v1alpha1.admissionregistration.k8s.io v1alpha1.internal.apiserver.k8s.io v1alpha1.storage.k8s.io v1alpha2.coordination.k8s.io v1alpha3.resource.k8s.io v1beta1.admissionregistration.k8s.io v1beta1.networking.k8s.io v1beta1.resource.k8s.io v1beta1.storage.k8s.io v2.autoscaling]
  	Type:               "Dangling",
- 	Status:             "",
+ 	Status:             "False",
- 	LastTransitionTime: v1.Time{},
+ 	LastTransitionTime: v1.Time{Time: s"2025-01-21 19:59:28.22697537 +0000 UTC m=+2.405278625"},
- 	Reason:             "",
+ 	Reason:             "Found",
- 	Message:            "",
+ 	Message:            `This FlowSchema references the PriorityLevelConfiguration object named "system" and it exists`,
  }
  	Type:               "Dangling",
- 	Status:             "",
+ 	Status:             "False",
- 	LastTransitionTime: v1.Time{},
+ 	LastTransitionTime: v1.Time{Time: s"2025-01-21 19:59:28.232527805 +0000 UTC m=+2.410831090"},
- 	Reason:             "",
+ 	Reason:             "Found",
- 	Message:            "",
+ 	Message:            `This FlowSchema references the PriorityLevelConfiguration object named "node-high" and it exists`,
  }
  	Type:               "Dangling",
- 	Status:             "",
+ 	Status:             "True",
- 	LastTransitionTime: v1.Time{},
+ 	LastTransitionTime: v1.Time{Time: s"2025-01-21 19:59:28.236808066 +0000 UTC m=+2.415111321"},
- 	Reason:             "",
+ 	Reason:             "NotFound",
- 	Message:            "",
+ 	Message:            `This FlowSchema references the PriorityLevelConfiguration object named "exempt" but there is no such object`,
  }
  	Type:               "Dangling",
- 	Status:             "",
+ 	Status:             "False",
- 	LastTransitionTime: v1.Time{},
+ 	LastTransitionTime: v1.Time{Time: s"2025-01-21 19:59:28.236810496 +0000 UTC m=+2.415113761"},
- 	Reason:             "",
+ 	Reason:             "Found",
- 	Message:            "",
+ 	Message:            `This FlowSchema references the PriorityLevelConfiguration object named "leader-election" and it exists`,
  }
  	Type:               "Dangling",
- 	Status:             "",
+ 	Status:             "False",
- 	LastTransitionTime: v1.Time{},
+ 	LastTransitionTime: v1.Time{Time: s"2025-01-21 19:59:28.244349304 +0000 UTC m=+2.422652559"},
- 	Reason:             "",
+ 	Reason:             "Found",
- 	Message:            "",
+ 	Message:            `This FlowSchema references the PriorityLevelConfiguration object named "leader-election" and it exists`,
  }
  	Type:               "Dangling",
- 	Status:             "",
+ 	Status:             "False",
- 	LastTransitionTime: v1.Time{},
+ 	LastTransitionTime: v1.Time{Time: s"2025-01-21 19:59:28.244351044 +0000 UTC m=+2.422654299"},
- 	Reason:             "",
+ 	Reason:             "Found",
- 	Message:            "",
+ 	Message:            `This FlowSchema references the PriorityLevelConfiguration object named "workload-high" and it exists`,
  }
[-]poststarthook/scheduling/bootstrap-system-priority-classes failed: not finished
[-]poststarthook/priority-and-fairness-config-producer failed: not finished
  	Type:               "Dangling",
- 	Status:             "",
+ 	Status:             "False",
- 	LastTransitionTime: v1.Time{},
+ 	LastTransitionTime: v1.Time{Time: s"2025-01-21 19:59:28.253019695 +0000 UTC m=+2.431322950"},
- 	Reason:             "",
+ 	Reason:             "Found",
- 	Message:            "",
+ 	Message:            `This FlowSchema references the PriorityLevelConfiguration object named "workload-high" and it exists`,
  }
  	Type:               "Dangling",
- 	Status:             "",
+ 	Status:             "False",
- 	LastTransitionTime: v1.Time{},
+ 	LastTransitionTime: v1.Time{Time: s"2025-01-21 19:59:28.253021215 +0000 UTC m=+2.431324470"},
- 	Reason:             "",
+ 	Reason:             "Found",
- 	Message:            "",
+ 	Message:            `This FlowSchema references the PriorityLevelConfiguration object named "workload-high" and it exists`,
  }
  	Type:               "Dangling",
- 	Status:             "",
+ 	Status:             "False",
- 	LastTransitionTime: v1.Time{},
+ 	LastTransitionTime: v1.Time{Time: s"2025-01-21 19:59:28.253025715 +0000 UTC m=+2.431328970"},
- 	Reason:             "",
+ 	Reason:             "Found",
- 	Message:            "",
+ 	Message:            `This FlowSchema references the PriorityLevelConfiguration object named "workload-high" and it exists`,
  }
  	Type:               "Dangling",
- 	Status:             "",
+ 	Status:             "False",
- 	LastTransitionTime: v1.Time{},
+ 	LastTransitionTime: v1.Time{Time: s"2025-01-21 19:59:28.266023747 +0000 UTC m=+2.444327022"},
- 	Reason:             "",
+ 	Reason:             "Found",
- 	Message:            "",
+ 	Message:            `This FlowSchema references the PriorityLevelConfiguration object named "exempt" and it exists`,
  }
  	Type:               "Dangling",
- 	Status:             "",
+ 	Status:             "False",
- 	LastTransitionTime: v1.Time{},
+ 	LastTransitionTime: v1.Time{Time: s"2025-01-21 19:59:28.266032617 +0000 UTC m=+2.444335882"},
- 	Reason:             "",
+ 	Reason:             "Found",
- 	Message:            "",
+ 	Message:            `This FlowSchema references the PriorityLevelConfiguration object named "global-default" and it exists`,
  }
  	Type:               "Dangling",
- 	Status:             "",
+ 	Status:             "False",
- 	LastTransitionTime: v1.Time{},
+ 	LastTransitionTime: v1.Time{Time: s"2025-01-21 19:59:28.266040247 +0000 UTC m=+2.444343502"},
- 	Reason:             "",
+ 	Reason:             "Found",
- 	Message:            "",
+ 	Message:            `This FlowSchema references the PriorityLevelConfiguration object named "workload-low" and it exists`,
  }
  	Type:               "Dangling",
- 	Status:             "True",
+ 	Status:             "False",
- 	LastTransitionTime: v1.Time{Time: s"2025-01-21 19:59:28 +0000 UTC"},
+ 	LastTransitionTime: v1.Time{Time: s"2025-01-21 19:59:28.266050447 +0000 UTC m=+2.444353702"},
- 	Reason:             "NotFound",
+ 	Reason:             "Found",
  	Message: strings.Join({
  		"This FlowSchema references the PriorityLevelConfiguration object",
  		` named "exempt" `,
- 		"but there is no such object",
+ 		"and it exists",
  	}, ""),
  }
  	Type:               "Dangling",
- 	Status:             "",
+ 	Status:             "False",
- 	LastTransitionTime: v1.Time{},
+ 	LastTransitionTime: v1.Time{Time: s"2025-01-21 19:59:28.284227301 +0000 UTC m=+2.462530576"},
- 	Reason:             "",
+ 	Reason:             "Found",
- 	Message:            "",
+ 	Message:            `This FlowSchema references the PriorityLevelConfiguration object named "catch-all" and it exists`,
  }
[-]poststarthook/scheduling/bootstrap-system-priority-classes failed: not finished
[-]poststarthook/scheduling/bootstrap-system-priority-classes failed: not finished
[-]poststarthook/scheduling/bootstrap-system-priority-classes failed: not finished
[-]poststarthook/scheduling/bootstrap-system-priority-classes failed: not finished
[-]poststarthook/scheduling/bootstrap-system-priority-classes failed: not finished
[-]poststarthook/scheduling/bootstrap-system-priority-classes failed: not finished
[-]poststarthook/scheduling/bootstrap-system-priority-classes failed: not finished
[-]poststarthook/scheduling/bootstrap-system-priority-classes failed: not finished
[-]poststarthook/scheduling/bootstrap-system-priority-classes failed: not finished
[-]poststarthook/scheduling/bootstrap-system-priority-classes failed: not finished
[-]poststarthook/scheduling/bootstrap-system-priority-classes failed: not finished
[-]poststarthook/scheduling/bootstrap-system-priority-classes failed: not finished
[-]poststarthook/scheduling/bootstrap-system-priority-classes failed: not finished
[-]poststarthook/scheduling/bootstrap-system-priority-classes failed: not finished
[-]poststarthook/scheduling/bootstrap-system-priority-classes failed: not finished
leases.coordination.k8s.io "leader-election-controller" not found
leases.coordination.k8s.io "leader-election-controller" not found
    testserver.go:582: Resolved testserver package path to: "/home/prow/go/src/k8s.io/kubernetes/cmd/kube-apiserver/app/testing"
    testserver.go:402: runtime-config=map[api/all:true]
    testserver.go:403: Starting kube-apiserver on port 42539...
    testserver.go:438: Waiting for /healthz to be ok...
[-]poststarthook/start-apiextensions-controllers failed: not finished
[-]poststarthook/crd-informer-synced failed: not finished
[-]poststarthook/start-service-ip-repair-controllers failed: not finished
[-]poststarthook/scheduling/bootstrap-system-priority-classes failed: not finished
[-]poststarthook/priority-and-fairness-config-producer failed: not finished
[-]poststarthook/bootstrap-controller failed: not finished
[-]poststarthook/apiservice-registration-controller failed: not finished
[-]poststarthook/apiservice-discovery-controller failed: not finished
[-]autoregister-completion failed: missing APIService: [v1. v1.admissionregistration.k8s.io v1.apiextensions.k8s.io v1.apps v1.authentication.k8s.io v1.authorization.k8s.io v1.autoscaling v1.batch v1.certificates.k8s.io v1.coordination.k8s.io v1.discovery.k8s.io v1.events.k8s.io v1.flowcontrol.apiserver.k8s.io v1.networking.k8s.io v1.node.k8s.io v1.policy v1.rbac.authorization.k8s.io v1.scheduling.k8s.io v1.storage.k8s.io v1alpha1.admissionregistration.k8s.io v1alpha1.internal.apiserver.k8s.io v1alpha1.storage.k8s.io v1alpha2.coordination.k8s.io v1alpha3.resource.k8s.io v1beta1.admissionregistration.k8s.io v1beta1.networking.k8s.io v1beta1.resource.k8s.io v1beta1.storage.k8s.io v2.autoscaling]
[-]poststarthook/scheduling/bootstrap-system-priority-classes failed: not finished
[-]poststarthook/priority-and-fairness-config-producer failed: not finished
[-]poststarthook/bootstrap-controller failed: not finished
[-]poststarthook/apiservice-registration-controller failed: not finished
[-]poststarthook/apiservice-discovery-controller failed: not finished
[-]poststarthook/scheduling/bootstrap-system-priority-classes failed: not finished
[-]poststarthook/priority-and-fairness-config-producer failed: not finished
[-]poststarthook/apiservice-registration-controller failed: not finished
[-]poststarthook/apiservice-discovery-controller failed: not finished
[-]poststarthook/scheduling/bootstrap-system-priority-classes failed: not finished
[-]poststarthook/scheduling/bootstrap-system-priority-classes failed: not finished
[-]poststarthook/scheduling/bootstrap-system-priority-classes failed: not finished
[-]poststarthook/scheduling/bootstrap-system-priority-classes failed: not finished
[-]poststarthook/scheduling/bootstrap-system-priority-classes failed: not finished
[-]poststarthook/scheduling/bootstrap-system-priority-classes failed: not finished
[-]poststarthook/scheduling/bootstrap-system-priority-classes failed: not finished
[-]poststarthook/scheduling/bootstrap-system-priority-classes failed: not finished
[-]poststarthook/scheduling/bootstrap-system-priority-classes failed: not finished
[-]poststarthook/scheduling/bootstrap-system-priority-classes failed: not finished
[-]poststarthook/scheduling/bootstrap-system-priority-classes failed: not finished
[-]poststarthook/scheduling/bootstrap-system-priority-classes failed: not finished
[-]poststarthook/scheduling/bootstrap-system-priority-classes failed: not finished
[-]poststarthook/scheduling/bootstrap-system-priority-classes failed: not finished
[-]poststarthook/scheduling/bootstrap-system-priority-classes failed: not finished
[-]poststarthook/scheduling/bootstrap-system-priority-classes failed: not finished
[-]poststarthook/scheduling/bootstrap-system-priority-classes failed: not finished
client rate limiter Wait returned an error: context deadline exceeded
    leaderelection_test.go:162: Expected the cle lease lock to transition to the first apiserver
--- FAIL: TestCoordinatedLeaderElectionLeaseTransfer (25.47s)
}
```

### Anything else we need to know?

N/A

### Relevant SIG(s)

/sig api-machinery

--- a/test/integration/apiserver/coordinatedleaderelection/leaderelection_test.go
+++ b/test/integration/apiserver/coordinatedleaderelection/leaderelection_test.go
@@ -41,7 +41,13 @@ func TestCoordinatedLeaderElectionLeaseTransfer(t *testing.T) {
 	timers := leaderelection.LeaderElectionTimers{
 		LeaseDuration: 5 * time.Second,
 		RenewDeadline: 3 * time.Second,
-		RetryPeriod:   2 * time.Second,
+		// RetryPeriod is intentionally set low because integration tests
+		// have a 10s timeout limit for wait.Poll(...) operations.
+		// This test forces an apiserver to give up its lease and enter a state of
+		// backoff when attempting to acquire the lease. Given the default JitterFactor of 1.2
+		// the maximum delay in renewal is 1 * (1 + 1.2) = 2.2s
+		// providing multiple renewal chances to minimize test flake.
+		RetryPeriod: 1 * time.Second,
 	}
 
 	featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, genericfeatures.CoordinatedLeaderElection, true)
@@ -134,6 +140,20 @@ func TestCoordinatedLeaderElectionLeaseTransfer(t *testing.T) {
 
 	// Shutdown the second apiserver
 	server2.TearDownFn()
+
+	// Forcefully expire the lease so the transition will be faster. Waiting the full duration could cause flakes.
+	// This must be done before the VAP on the first apiserver is removed to avoid conflicts in updating.
+	lease, err = clientset.CoordinationV1().Leases("kube-system").Get(ctx, "leader-election-controller", metav1.GetOptions{})
+	if err != nil {
+		t.Fatal(err)
+	}
+	if lease.Spec.HolderIdentity != nil && *lease.Spec.HolderIdentity != leaseName {
+		lease.Spec.RenewTime = &metav1.MicroTime{Time: time.Now().Add(-30 * time.Second)}
+		_, err = clientset.CoordinationV1().Leases("kube-system").Update(ctx, lease, metav1.UpdateOptions{})
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
 
 	// Allow writes again from the first apiserver
 	err = clientset.AdmissionregistrationV1().ValidatingAdmissionPolicies().Delete(ctx, vap.Name, metav1.DeleteOptions{})

